# Welcome Back to Another Exciting Week

# Annoucement:

1. Lab 6 released - Due 04/08 at 8 PM!

# Please remember

1. Submit on Jupyter
2. Upload to Gradescope
3. Assign questions to pages on gradescope for those that we ask about

If you have questions, please ask!

# Today's Agenda:

1. Functions
2. P - values
3. Hypothesis testing
4. Estimating P-Values with a simulation!

# Important - Lab 6 is more involved. Please take your time.

# Also email me with questions or for office hours!

# 1. Functions!

Functions are basically the last BIG thing we haven't covered. Once you have functions, you have all the must fundamental building blocks to basically start programming. Congratulations.

Think of functions like a formula. You create a function, and when you use it you give it inputs.

Python will take those inputs and do the same thing to them, just like a formula.

This means that if you have anything vaguely repetitive, you can make it a function and only need to write the code once.

It also allows you to break up you code. Now, each part of your code can be done by different functions. So when things break, you only need to fix one thing.

Let's see an example.

What about a simple function that adds two numbers:

In [2]:
def add_function(a,b):
    c = a + b
    return c

In [3]:
add_function(12,17)

29

Ok, so what is going on.

You first create functions with `def`. 

After that is the name of the function, this cans `add_function` (you cannot use spaces).

Then, there are parenthesis. Inside, you put the names of the arguements you want. In this case we have `a` and `b`. These are placeholders. You can use them, as if they exist, when writing the function.

But, when the time comes you put whatever you want in for those placeholders!

Then, the function is doing something, in this case adding `a` and `b` together, and creating a new local variable, `c` which is assigned to that sum.

Finally, `return` is how we end functions. `return` is what tells the function to output. 

We are outputting `c` this time. Importantly, once a function runs `return` the function is over. 

Also, notice the indentation. Just like with If and for loops, we have to indent everything in the function over to be under it.


Do `a`, `b`, or `c` exist now?

In [4]:
print(c)

NameError: name 'c' is not defined

NOPE! They are what is called `locals`. What that means is that they are created when the function runs, and then deleted afterwards. 

The variables we create ourselves that persist are called `globals`.

Let's see a more complicated example.

In [7]:
import numpy as np

def pythag(a,b):
    return np.sqrt(a**2 + b**2)

In [8]:
pythag(3,4)

5.0

Notice, I could have made a third variable, and then returned that, but in this case I can also just return the output of that statement.

I can return anything I want, and I can also print within a function.

In [9]:
def is_string(my_str):
    if type(my_str) == str:
        print("its a string!")
    else:
        print("bummer, not a string")
        
    

In [10]:
is_string("this is a string dude")

its a string!


In [12]:
is_string(5)

bummer, not a string


Notice, there isn't even a return statement!

I can also have multiple return statements, BUUUUUUT... only one of them will ever execute when I run the function.

In [13]:
def is_two(a):
    if a == 2:
        return True
    else:
        return False

In [14]:
is_two(2)

True

In [15]:
is_two(5)

False

In [16]:
is_two('two')

False

The above function ends as soon as it hits a return statement, but in those cases where the number is not 2, then it never runs the first return statement.

Finally, lets do the quadratic formula.

In [17]:
def quad(a,b,c):
    
    print("Calculating Determinate")
    
    
    if b**2 - 4*a*c >= 0:
        determinant = np.sqrt(b**2 - 4*a*c)
        print("Determinate is real, Calculating answer")
        answer_pos = (-b + determinant)/2*a
        answer_neg = (-b - determinant)/2*a
        
        print("Anwer is:")
        return answer_pos, answer_neg
    
    else: 
        print("No real solution pal")
    
    
    

In [18]:
quad(1,4,4)

Calculating Determinate
Determinate is real, Calculating answer
Anwer is:


(-2.0, -2.0)

In [19]:
quad(1,2,4)

Calculating Determinate
No real solution pal


In [20]:
quad(1,4,2)

Calculating Determinate
Determinate is real, Calculating answer
Anwer is:


(-0.5857864376269049, -3.414213562373095)

In [23]:
answer, answer_two = quad(1,4,2)

Calculating Determinate
Determinate is real, Calculating answer
Anwer is:


In [24]:
answer

-0.5857864376269049

In [25]:
answer_two

-3.414213562373095

# 2. P - Values

What are they?

With me now: P - values are the probability of viewing this data given the hypothesis!

P-values <b> ARE NOT </b> the probabity the hypothesis is true!

P-values are the probability of the data given the hypothesis.

<b> THEY ARE NOT the probability of the hypothesis given the data! </b>

What does this mean?

Well, what is the probability of rolling 1 in on a six sided dice? 

It is 1 in 6, or 1/6, or .16666.  

So if I rolled a dice and got 1, and my hypothesis is that the dice is fair, then the p-value is 1/6 or .166.

Ok, so what if I rolled a dice twice and got one both times. We'll the probability of that is 1/6 * 1/6. 

Why, because they are independent - meaning what I rolled the first time doesn't affect my second roll.

When that is true, we can just multiple the probability of two independent events (in this case two dice rolls) together to get the joint probability (in this case the probability of getting two rolls).

So the odds of a 1 and then 1 one are just (1/6)**2 or 1/36 or .0277

Ok, so lets say I rolled this dice six times instead of two, and I got 1 each time. The p-value for that is:

.00002143

REALLY Unlikely. 

So what is the probability that the null hypothesis, that this dice is fair, is true?

<b> WE DO NOT KNOW!!!!! </b>

BUT, we do know that the data we observe is VERY unlikey to occur if the null is true. We know that it only occurs .002143 percent of the time (notice we lost two zero because we are in percent).

Basically, the P-value says that if the dice is fair, the probability of seeing six ones in a row is .00002143. 

So what do we learn?

Well, this is science at its weirdest. 

We don't know the probability the null (this dice is fair) is true. 

But we do know all the data we observe is really, really unlikely if the null is true. 

So, we can reject the null. 

Basically, we say this is so unlikely to occur given if the null was in fact true, that we reject the null...its got to be something else.

P-values are very, very important to scientists. They are not complicated - but many people, scientists included, get them wrong. 

Please, please, please, please spend some time with them, think about it carefully.

And please email if you want to talk about them more.




# 3. Statistical Significance & Hypothesis Testing

So what does it mean what a finding is "statistically significant"?

Many scientific studies, particularly empirical studies, test a hypothesis.

## The Null Hypothesis

What this means is the researchers test what is a called a <b> "null" </b> hypothesis.

Think of the null as a baseline. Is it the starting point.

For example, if I want to test if there is a relationship between smoking and cancer, the null hypothesis is that there is no relation ship.

The null is the starting point. It is conservative, in that when we are testing to see if there is a relationship between two things, we almost always start by assuming there is no relationship.

## The Alternative Hypothesis

The <b> "alterative" </b> hypothesis is that the null is not true. 

For example, if our null is that there is no relationship between smoking and cancer, our alternative hypothesisis that there is a relationship between smoking and cancer. 

So our null is always conservative, in the sense that we assume (as our baseline) in the null, that what we think might be there is not there.


## Testing the Null

We then do an experiment, or use data, and test the null. If we find no strong evidence against the null then we "fail to reject the null".

Conversely, if we find strong evidence against the null, we "reject the null". We do not accept the alternative - because we don't know that it is true, we just have strong evidence the null isn't true.


## Statistical Significance

So what is statistical significance? 

Statistical significance is the criteria we either meet, and reject the null, or we do not meet, and then we fail to reject the null.

Basically, statistical significance is the rule we use to decide if we are going to reject or fail to reject null.

What is this rule?

In practice, we use the P-value!

Remember, the P-value assumes a hypothesis. In this case, we assume the null hypothesis.

We then see how likely the data, or something more extreme, is likely to occur, if we assume the null.

If the P-value is below a certain threshold, we reject the null, otherwise we fail to reject it.

So statistical significance is determined by the threshold that the field (economics, political science, physics, etc) has decided is the threshold. 

Statistical significance is just a way of formally saying "the P-value is below our threshold, so we can reject the null hypothesis".

Essentially, the data is so unlikely to occur given the null hypothesis, that we reject the null. 

Statistical significance is just formally identifying what the threshold is for unlikely.

For much of the social sciences, that value is .05 or 5%.

Which means that if we get a P-value, given a null, below .05, then we can reject the null. 

It also means that we know we will reject the null sometimes, EVEN WHEN THE NULL IS TRUE!

## Statistical Significance is not Perfect!

Why? Because we know, if the null is true, that sometimes weird/rare thing things do happen.

Data that only occurs under the null 5% of the time or less (P-vale of .05 or less) does actually occur, its just rare.

But we will reject the null if the P-value is that low, even if the null is true and the data we have is just weird.

Here is an example. The average height of men in the U.S. is about 5'10'' (that is five feet ten inches). Let us say that 5'10'' is our null for height.

Now let's say we take a random sample, and just by weird random luck, we only sample basketball players. And our sample of 20 has an average height of 6'5'' (really tall). 

The chances of that are incredibly small! But it can happen. So this case, the P-value of having a sample of 20, whose height is 7 inches away from the mean might be really, really low, like .0001 or .01%. 

In this case we reject the null. But we know the truth - average male height really is 5'10'' - the null is true. But if we got a weird sample, just by bad luck, we would reject the null. 

<b> This is why replication is critical in science </b>

Sometimes we find weird things just because we got a weird sample.

But if a bunch of studies keep doing things with different samples, it is really unlikely for them to consistently get weird samples.

As we do more studies, we are more and more likely to discover the truth.

So - as you know - one study never proves anything!

# 4. Estimating P-Values with a simulation!

Sometimes it is really easy to calculate a P-value directly (what is also called analytically).

In this case, we know all the probabilities associated with the null (they are all 1/6), so we can calculate the probability for given roll. Easy.

But sometimes it is just really hard to calcuate how likely the data is to occur under the null.

So instead, we turn to simulation.

In a simulation, we ASSUME the null is true, and create fake data.

We then look at the fake data created by our simulation. 

We can then count how many times the same numbers in the real data occur in the fake data, which gives us an estimated P-value.

Basically, we simulate a fake world where the null hypothesis is true. We then see how often what actually happened (the real data) occurs in the fake world. 

If what actually happened (real data) is very rare in the simulation, then we the assumptions of the simulation (the null hypothesis) can be rejected.

If what we see in the real world (real data) is very common in the simulation, then we cannot reject the null hypothesis - basically because what actually happened seems to be very common under the null according to our simulation.

Let's see an example.

# Coin Simulation

Let's do a very simple coin simulation.

We have a coin, and we want to test if it is fair. 

Basically, we are going to assume the coin is fair (it is our null) because almost all coins are fair.

Our alternative hypothesis is that the coin is not fair.

Let's say someone flipped the coin 10 times, and got 9 heads and one 1 tail. 

So, we will simulate flipping a fair coin 10 times - but we will do this 10,000 times.

And then, of those 10,000 simulations, we will see how many times we got 9 or 10 heads.

In this case, the P-value is the probability of getting what we got (9 heads) or something more extreme (10 heads). 

Basically, how likely is it to get 9 or more heads if the coin is fair?

If the P-value is less than .05, we will reject the null and say that this coin is not fair. 


In [27]:
np.random.seed(0) #Ignore this for now - it is not necessary but it helps me 

num_sim = 10000 #number of simulations

num_flips = 10 #number of coin flips in a simulation

coin_prob = .5 #the null hypothesis - in this case the coin is fair.

count_coin = 0 #this will keep track of all the number times we get 9 or 10 heads

for a_sim in range(num_sim):
    
    flip_list = 0 #this keeps track of all the heads in a given simulation (will always be between 0 and 10)
    
    for flip_sim in range(num_flips): #this is the loop where we will simulate 10 flips.
        
        if np.random.rand() < coin_prob: #if our coin is heads
            
            flip_list += 1 #then count 1 for a heads
            
            
    if flip_list >= 9: #after we do 10 flips, if 9 or 10 of them are heads
        
            count_coin += 1 #Then count that
            


The function `np.random.rand()` generates a random number between 0 and 1 UNIFORMLY.

So if `coin_prob` is .5, half the time `np.random.rand()` will generate a number above .5, and half the time below it. So this is how we simulate flipping a coin on a computer.

Ok, so now, we know we did 10,000 simulations, and `count_coin` has added 1 for every single time we got a 9 or 10 (it basically keeping track of how many times what actually happened occured in our simulation).

So, how many times did it occur?

That is just the number of times what actually occured over the number of total simulations

In [29]:
count_coin/num_sim

0.011

In [28]:
count_coin

110

Low huh. 

Basically, only 1% of time (P-value of .011), or 110 out of 10,000, did we get 9 or 10 heads in 10 flips.

The P-value is way below our threshold of .05 or 5%. 

Therefore we reject the null that the coin is fair. 

Basically, we are saying if you said you flipped a coin ten times and got 9 or more heads, I would reject the hypthesis that it was a fair coin.

BUT! We know it does happen! It happened 110 times - so it can happen.

But it is sufficiently rare that we still reject the null.

Basically, its too unlikely given the null - so we reject it. 

# Baseball Simulation

In the world series, two baseball teams play seven games. Whichever team wins four them (and since there are only two teams one team must always win at least four), wins the world series!

Now let's say the world Series just ended and the Yankees won and the Red Soxs lost. This is excellent.

Unfortunately, the redsox are very, very sore losers.

They are so sore, and because they know a little bit of statistics, they say:

"We think we are actually much better, and you just won by luck. In any given game there is a 60% chance we will win".

The Yankee fan, who is very smart and sophisticated about these things, says:
"Excellent null hypothesis. Let's simulate 10,000 world series where the Red Sox have a 60% change of winning each game, and see what percentage of the world series the Red Sox actually lose."

Basically, if we simulate 10,000 world series assuming the redsox win 60% of games (and remember there are 7 games in a series), and we find the redsox very, very rarely lose the series, it means that what actually happened (the red sox losing) is very unlikely to occur if the redsox fan is right.

So in this case, we will assume the null is that the redsox have a chance of winning each game of .6 (or 60%) and our alternative is that their chance if winning each game is not .6.

In order to do this, we need to build some coding infastructure.

Firstly, let's write a that can simulate a game.

This function will take a probability (in this case we will use .6) and return one if the redsox win, and 0 if they lose.


In [32]:
def soxwin(prob):
    if np.random.rand() < prob:
        return 1
    else: 
        return 0

Again, the function `np.random.rand()` generates a random number between 0 and 1 UNIFORMLY. 

So, if we say a = .6, then 60% of the time `np.random.rand()` will be less than that number and we will get a 1 (redsox win that game), and 40 `np.random.rand()` will be above that number and we will get 0 (redsox lose). 

If we change the a = .8, it will work exactly the same, but in 80% of simulates the redsox will win rather than 60%.

COOL.

Now, lets do TWO for loops.

First, one for 10,000 simulates of the world series.

THEN, for each world series, we simulate 7 games.

In [45]:
np.random.seed(0)

num_sims = 100000 #total number of world series

num_games = 7 # total number of games in a world series

sox_win_prob = .6 #probability the sox win a given game

win_count = 0 #this  will record now many times the red sox win.

for this_world_series in range(num_sims):
    
    a_world_series = 0 # this keeps track of the games in a single world series
    
    for this_game in range(num_games):
        
        a_world_series += soxwin(sox_win_prob) # every win is a 1, every loss is a 0
        
    if a_world_series >= 4: #if the redsox won 4 or more games in this simulation
        win_count += 1 #then add 1


Ok, so our list, `win_count` has a 1 for every win for the redsox (and zero for every loss).

SO, if we just sum it up, we can see how many world series wins they have, and then divide that by the number of simulations (in this case 10,000).

That gives us a P-value! 

It gives us the probability of the redsox winning the world series if their changes of winning any given game is 60%. 

It also gives us the P-value of them LOSING the world series given the null. 

Let's see.

In [46]:
win_count

97085

In [47]:
win_count/num_sims

0.97085

So, roughly 71% of the time the redsox win the world series if they win 60% of games. Importantly, our estimate it rough because we simulated it (but we did a lot so it must be close by the LLN!).

It also means, on average, they would only lose 29% of the time (1 - .71). 

So the P-value here is .29, because what we saw in real life was the redsox losing.

So under the null that the redsox win 60% of their games, the p-value of them losing the world series (as they did) is .29.

In this case, we fail to reject the null!

It is more likely than not that if the redsox win 60% of games then they will lose the world series (they win it 71% of time).

BUT, we don't have strong enough evidence that the null is wrong.

It is very, very possible for them to win 60% of games and still lose the world series.

Now, what is the P-value of them losing if the null was that they win 75% of their games against the yankees?

You only need to change one number above to check that!

Even more interesting, let's assume the redsox really do win 60% of their games against the yankees. 

How many games would world series need to have (remember in reality it is seven), for the redsox to win 99% percent of the time?